matcher.go

```diff
@@ -369,6 +310,100 @@ func (m *Matcher) matchExec(token tokenType, tokenData []byte, node *ExecNode) e
 	return nil
 }
 
+// Returns an error code, and a boolean to dictate whether or not for the caller to return immediately
+func (m *Matcher) matchObjectOrArray(token tokenType, tokenData []byte, node *ExecNode) (error, bool) {
+	var keyLitParse fastLitParser
+	var endToken tokenType
+	var arrayIndex int
+	var arrayMode bool
+
+	switch token {
+	case tknObjectStart:
+		endToken = tknObjectEnd
+	case tknArrayStart:
+		endToken = tknArrayEnd
+		arrayMode = true
+	default:
+		panic("Unexpected type input for function call matchObjectOrArray")
+	}
+
+	for i := 0; ; i++ {
+		// If this is not the first entry in the object, there should be a
+		// list delimiter ('c') that shows up in the input first.
+		if i != 0 {
+			token, _, err := m.tokens.Step()
+			if err != nil {
+				return err, true
+			}
+
+			switch token {
+			case endToken:
+				return nil, true
+			case tknListDelim:
```

Comment:
should we verify arrayMode = true?

Reply:
I think that this logic will need to be improved to performantly handle both arrays and objects while still enabling us to do various kinds of indexings.

Reply:
I converted the original if statements into this switch statement and added a new case. I'm not sure what @brett19 you mean by performantly handle?

Reply:
@ysui6888 Good point you raised, but we shouldn't verify, because from the comment above, it's possible to see list delimiter before an object